### PR TITLE
[Metricbeat][Kubernetes] Update state_* metricsets to be compatible with supported kubernetes versions

### PR DESCRIPTION
**Pull request by constanca-m**
**This PR is going to be split. Please see https://github.com/elastic/beats/pull/34432.**

## What does this PR do?
Update the metrics of the `state_*` metricsets.

**WARNING**: Any dashboards that use these metrics were not looked into in this PR; `state_container` and `state_resourcequota` metricsets functions were not modified, unlike the others.

## Why is it important?
Kube state metrics versions tested before are no longer supported by the metricbeat Kubernetes module.

Changes:

**New files**:
- Updated the `data.json` examples
- Deleted the previous files for KSM, and added new ones that support versions 1.23.x-1.26.x of K8s
- Added `state_metricset.go` file with all the common functions between the `state_*` metricsets (expect for resource quota and container, as they don't follow the same logic as the others)
- Added new file `ktest.go` for kubernetes related testing functions

**New test**:
`TestStateMetricsFamily` was added to all testing files of the `state_*` metricsets.
The only purpose of it, is to help us remove deprecated metrics. It simply goes over the kube state metrics files, and checks if the family metric exists in it. It does **not** check for labels, as they may not appear in the files, even though they may exist - e.g. if there is no data for that metric.

**Updated file**:
`metricbeat/module/kubernetes/_meta/test/docs/README.md` was changed to make it more clear for us to know how to test the metrics.

Most of the **changes on the metrics/labels** are:

- State container:
   - `kube_pod_container_resource_limits_*_cores` replaced by `kube_pod_container_resource_limits`
   - `kube_pod_container_resource_requests_*_cores` replaced by `kube_pod_container_resource_requests`
   - `kube_pod_container_status_terminated` removed since it can be easily obtained through `kube_pod_container_status_terminated_reason`
   - `kube_pod_container_status_waiting` removed since it can be easily obtained through `kube_pod_container_status_waiting_reason`
   - `kube_pod_container_status_restarts` removed - it does not have data in any of the KSM versions supported by us
   - `reason` added - from `kube_pod_container_status_*_reason`
- State job:
  - `condition` label added
- State node:
  - `kube_node_status_allocatable_*` replaced by `kube_node_status_allocatable`
  - `kube_node_status_capacity_cpu_cores_*` replaced by `kube_node_status_capacity`
  - `condition` label added
  - `status` label added
  - `resource` label added
  - `unit` label added
- State storage class:
  - `volumeBindingMode` removed
  - `reclaimPolicy` removed

You can find more information about the metrics in the README file of each metricset.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Relates to https://github.com/elastic/beats/pull/34275

## How to test
Please, see [this file](https://github.com/constanca-m/beats/blob/update-kube-state-metrics-tests/metricbeat/module/kubernetes/_meta/test/docs/README.md).

### CODE REVIEW

**Comment by gizas** [comment]
@constanca-m can we have 2 additions in the description of the PR:
- A section `How to Test`
- A section like `Outcome` with proof of successful run of test cases? Maybe the bash output or some screenshots?

**Comment by constanca-m** [comment]
> @constanca-m can we have 2 additions in the description of the PR:
> 
>     * A section `How to Test`
> 
>     * A section like `Outcome` with proof of successful run of test cases? Maybe the bash output or some screenshots?

Yes @gizas . I added one, and I want to update the markdown files, but I was having trouble testing so many changes, that's why I opened the PR. But I still want to change more things

File: metricbeat/module/kubernetes/_meta/test/docs/README.md [hunk 1]
```diff
@@ -1,88 +1,103 @@
-# Testing Metricbeat
+## Testing Metricbeat
```
Comment (gsantoro) [inline metricbeat/module/kubernetes/_meta/test/docs/README.md:1]:
```suggestion
# Testing Metricbeat
```

File: metricbeat/module/kubernetes/_meta/test/docs/README.md [hunk 2]
```diff
@@ -1,88 +1,103 @@
...
 
+### Prerequisites
```
Comment (gsantoro) [inline metricbeat/module/kubernetes/_meta/test/docs/README.md:3]:
```suggestion
## Prerequisites
```

File: metricbeat/module/kubernetes/_meta/test/docs/README.md [hunk 3]
```diff
@@ -1,88 +1,103 @@
...
-kubectl port-forward deployment/kibana 5601:5601
-```
-
-
-## Playground Metricbeat Pod
+go test -data
+ ```
+
+
+### Deploy metricbeat
```
Comment (gsantoro) [inline metricbeat/module/kubernetes/_meta/test/docs/README.md:16]: couple of notes about this doc:
- it is not very clear if `Deploy metricbeat` is required for the `go test` or the integration tests or the cloud deployments. We could add a few words making sure that it is clear that those are 4 different way to test metricbeat and they are independent from each other. 1. local go test, 2. manual testing with `Deploy metricbeat`, 3. integration tests, 4. tests in cloud
- given that we have 9 steps only for `deploy metricbeat`, it would be great to have a taskfile to provide a single command to the user
-

File: metricbeat/module/kubernetes/_meta/test/docs/README.md [hunk 4]
```diff
@@ -1,88 +1,103 @@
...
+5. Deploy manifests:
+   - Testing `state_*` metricsets? Run:
+   ```bash
+   kubectl apply -k with-ksm
+    ```
+   - **Not** testing `state_*` metricsets? Run:
+   ```bash
+   kubectl apply -k without-ksm
+    ```
+   > **Note**: Adjust hosts for elasticsearch and kibana if they are not correct.
```
Comment (gsantoro) [inline metricbeat/module/kubernetes/_meta/test/docs/README.md:48]: I think this comment is misleading. I don't see why those hostnames should not be correct if they followed all the steps until now.

File: metricbeat/module/kubernetes/_meta/test/docs/README.md [hunk 5]
```diff
@@ -1,88 +1,103 @@
...
+
+6. Go to metricbeat directory and build the metricbeat binary:
+    ```bash
+    export CGO_ENABLED=0 && GOOS=linux GOARCH=amd64 go build
+    ```
+
+7. Copy the metricbeat binary:
+    ```bash
+   kubectl cp ./metricbeat `kubectl get pod -n kube-system -l k8s-app=metricbeat -o jsonpath='{.items[].metadata.name}'`:/usr/share/metricbeat/ -n kube-system
```
Comment (gsantoro) [inline metricbeat/module/kubernetes/_meta/test/docs/README.md:57]: Given how many steps are required and this workflow doesn't support `hot reload/rebuild` at code changes, I am wondering if we could use a simplified version of this Tiltfile https://github.com/elastic/beats/blob/main/dev-tools/kubernetes/Tiltfile that would include in a single command all the 9 steps and support `hot reload/rebuild` as well.

File: metricbeat/module/kubernetes/_meta/test/docs/README.md [hunk 6]
```diff
@@ -1,88 +1,103 @@
...
+9. Once inside the pod, run:
+   ```bash
+   metricbeat -e -c /etc/metricbeat.yml
+    ```
+
+> **Note**: In case you want to test an updated metricbeat binary, you should delete the manifests
+> and go back to step 5.
+
+
+# Running integration tests.
```
Comment (gsantoro) [inline metricbeat/module/kubernetes/_meta/test/docs/README.md:74]:
```suggestion
## Running integration tests.
```

File: metricbeat/module/kubernetes/_meta/test/docs/README.md (hunk 3 above)
Comment (gsantoro) [inline metricbeat/module/kubernetes/_meta/test/docs/README.md:16]:
```suggestion
## Deploy metricbeat
```

**Comment by gsantoro** [comment]
A couple of more comments (not code specific):
- since you use kustomize in this PR, I would personally add a link to this [issue](https://github.com/elastic/elastic-agent/issues/2175) so that we track this other use of kustomize
- as part of this [issue](https://github.com/elastic/beats/issues/33514) I was planning to add some overlays for metricbeat. In order to avoid overlapping our work, I suggest for you to keep the resources in the kustomization.yml as they are but we could just repoint the kustomization to the new overlays (that eventually will be located in the `deploy` folder) in the future.

**Review by gsantoro — CHANGES_REQUESTED** [review]
Can you please address those small changes. for the long term changes like Taskfile or/and Tiltfile I'm happy to discuss them together.

File: metricbeat/module/kubernetes/_meta/test/docs/README.md [hunk 7]
```diff
@@ -1,88 +1,103 @@
...
+    ```
+
+7. Copy the metricbeat binary:
+    ```bash
+   kubectl cp ./metricbeat `kubectl get pod -n kube-system -l k8s-app=metricbeat -o jsonpath='{.items[].metadata.name}'`:/usr/share/metricbeat/ -n kube-system
+    ```
+
+8. Get inside the pod:
+   ```bash
+   kubectl exec `kubectl get pod -n kube-system -l k8s-app=metricbeat -o jsonpath='{.items[].metadata.name}'` -n kube-system -it -- bash
```
[the remark this answers is not in the document]
Comment (constanca-m) [reply metricbeat/module/kubernetes/_meta/test/docs/README.md:62]: That is true, I didn't do that because I was planning on adding the comment to set the dashboards, but I forgot

File: metricbeat/module/kubernetes/_meta/test/docs/README.md (hunk 4 above)
Comment (constanca-m) [reply metricbeat/module/kubernetes/_meta/test/docs/README.md:48]: You are right, I will remove it

File: metricbeat/module/kubernetes/_meta/test/docs/without-ksm/metricbeat-kubernetes.yaml [hunk 8]
```diff
@@ -7,66 +7,29 @@ metadata:
   labels:
     k8s-app: metricbeat
 data:
-  metricbeat.yml: |-
+  metricbeat.yml: |-
```
Comment (gsantoro) [inline metricbeat/module/kubernetes/_meta/test/docs/without-ksm/metricbeat-kubernetes.yaml:10]: do you mind splitting this file into 1 file per resource. some of the advantages:
- easier to review if you touched a specific resource by looking at the name of the resource modified
- it is easier changing the order of the resource like at https://github.com/elastic/elastic-agent/pull/2104/files#diff-68eb49af89c50af014d3edcec563bc6ef6afbe3c099c5fb2bf177876bce92c72R1. I had a bug in the past that when deploying on Kind and the service account was defined last in the manifest it would take 6m to deploy all the resources compared to a couple of seconds if the resources were in right order (eg. service-account, roles, role-binding...)
- you can even refer to resources defined somewhere else (via a github link) that should be shared. For example you could avoid proving a manifest for a service account resource if already defined somewhere else. this should make it easier to follow the DRY principle

**Comment by MichaelKatsoulis** [comment]
@constanca-m I was initially confused by the PR title as I thought it only touches tests.
Could you update it to something like `Update state_* metricsets to be compatible with supported kubernetes versions` ?

I am also concerned about possible breaking changes. In the description you give a list of kube-state-metrics fields that changed in the newer versions. But we need also a list of the Elasticsearch fields(fields produced by the metricsets) that may change due to this PR. Fields that are deleted or added, I there are any.

**Comment by gsantoro** [comment]
@constanca-m I agree with @MichaelKatsoulis that the title is a bit misleading. For future reference, I would try to create smaller and more focused PRs if you can.

**Comment by constanca-m** [comment]
Ok @gsantoro and @MichaelKatsoulis , but I am a bit confused about that. Everything changed relates to the metrics of the updated KSM versions, so I don't think I could split this PR in more than one

**Comment by gsantoro** [comment]
off top of my head I see at least 3 PRs here (but they might be so dependent on each other that splitting them might be difficult):
- changes from the new kube-state-metrics version
- use kustomization for tests
- changes to k8s manifest (eg. alpine -> busybox) and other minor changes

**Comment by constanca-m** [comment]
> off top of my head I see at least 3 PRs here (but they might be so dependent on each other that splitting them might be difficult):
> 
>     * changes from the new kube-state-metrics version
> 
>     * use kustomization for tests
> 
>     * changes to k8s manifest (eg. alpine -> busybox) and other minor changes

For steps 2 and 3, I only did that to update the manual tests that we were using with the previous playground folder, but I had no intention in improving that

File: metricbeat/module/kubernetes/_meta/test/docs/README.md (hunk 5 above)
Comment (ChrsMark) [reply metricbeat/module/kubernetes/_meta/test/docs/README.md:62]: This would make Tilt the "official" way to go for anyone that wants to run Metricbeat from source, inside or outside the team (community devs). Do we want to add this extra requirment for all devs having to installing and maintaining Tilt in their dev environments? I have mixed feelings about this tbh. In the past tools that were introduced in general for Beats was an extra pain and were not maintained properly. @gizas what do you think here?

File: metricbeat/docs/fields.asciidoc [hunk 9]
```diff
@@ -45234,7 +45243,7 @@ type: boolean
 
 --
 
-*`kubernetes.container.status.restarts`*::
+*`kubernetes.container.status.restarts.count`*::
```
Comment (ChrsMark) [inline metricbeat/docs/fields.asciidoc:45246]: Do we completely remove fields? This would a breaking change. cc: @gizas

File: metricbeat/module/kubernetes/pod/pod.go [hunk 10]
```diff
@@ -44,7 +43,7 @@ var (
 // init registers the MetricSet with the central registry.
 // The New method will be called after the setup of the module and before starting to fetch data
 func init() {
-	mb.Registry.MustAddMetricSet("kubernetes", "pod", New,
+	mb.Registry.MustAddMetricSet("kubernetes", util.PodResource, New,
```
Comment (ChrsMark) [inline metricbeat/module/kubernetes/pod/pod.go:46]: That change (and similar) are irrelevant to what this PR's title indicate. Shall we change the PR tittle or split the PR into smaller ones?

File: metricbeat/module/kubernetes/_meta/test/docs/without-ksm/metricbeat-kubernetes.yaml [hunk 11]
```diff
@@ -187,6 +177,9 @@ spec:
               subPath: metricbeat.yml
             - name: data
               mountPath: /usr/share/metricbeat/data
+            - name: kibana
```
Comment (ChrsMark) [inline metricbeat/module/kubernetes/_meta/test/docs/without-ksm/metricbeat-kubernetes.yaml:180]: Out of curiosity, why is this needed?

**Review by ChrsMark — CHANGES_REQUESTED** [review]
In general the PR looks to be in the right direction. 
Some issue that we need to address:

1. I see in the description that some fields are removed, for example:

```
kube_pod_container_resource_limits_*_cores replaced by kube_pod_container_resource_limits
kube_pod_container_resource_requests_*_cores replaced by kube_pod_container_resource_requests
kube_pod_container_status_terminated removed since it can be easily obtained through
```
This is a breaking change and should be discussed if needed. In general we should not remove fields with no serious reason, since it would break users' dashboards, alerts, visualisations (even custom ones). Can we clarify this please here?

2. I see that golden files (.expected) are changed along with the code's changes. This means that unit-tests' files are changed at the same time with the code's changes. How we ensure here that code changes do not break anything? In principle that's the purpose of the unit tests to ensure that code changes do not break the functionality and do not break backwards compatibility. Are we confident here that code changes are transparent to the end results our users will face?
3. Are dashboards also tested? Since these changes touch the ingest part we need to make sure that pre-built dashboards are also functional.
4. In principle it's a good practice to split such PRs into multiple smaller ones with stanadlone changes so as to be easier to test and review and ensure that  those are easily reversible if sth goes south (same for handling their backports etc).


Only consider the first point as a blocker. The rest of them are just suggestions or ask for clarifications.

**Comment by MichaelKatsoulis** [comment]
@constanca-m see this [PR's](https://github.com/elastic/beats/pull/32037#issue-1279912861) description where controllermanager prometheus metrics where updated to take an idea of what both me and @ChrsMark ask.

**Comment by constanca-m** [comment]
>     1. I see in the description that some fields are removed, for example:
> 
> 
> ```
> kube_pod_container_resource_limits_*_cores replaced by kube_pod_container_resource_limits
> kube_pod_container_resource_requests_*_cores replaced by kube_pod_container_resource_requests
> kube_pod_container_status_terminated removed since it can be easily obtained through
> ```
> 
> This is a breaking change and should be discussed if needed. In general we should not remove fields with no serious reason, since it would break users' dashboards, alerts, visualisations (even custom ones). Can we clarify this please here?
 
>     2. I see that golden files (.expected) are changed along with the code's changes. This means that unit-tests' files are changed at the same time with the code's changes. How we ensure here that code changes do not break anything? In principle that's the purpose of the unit tests to ensure that code changes do not break the functionality and do not break backwards compatibility. Are we confident here that code changes are transparent to the end results our users will face?

Since we only support Kubernetes versions that are still being maintained, it doesn't really make sense to keep checking metrics from Kube state metrics versions that are no longer compatible to what we support. We have been removing these kind of files in previous [PRs](https://github.com/elastic/beats/pull/32037/files#diff-2275d963a655d60e65554a9e35bccb30697c9f6c394c9711451701844f9dcc44).

>     3. Are dashboards also tested? Since these changes touch the ingest part we need to make sure that pre-built dashboards are also functional.

They are not, I mentioned it in the PR description. I didn't check it because we still haven't established what we want to convey from them. I wanted to address that by writing [documentation in one of our resource metrics](https://github.com/elastic/observability-docs/issues/2474) (controller manager, in this case) so we could discuss what it is important and what is not. While we don't do that, I don't know if it is worth it changing so many dashboards to end up changing them later as they don't hold the meaning we want.

>     4. In principle it's a good practice to split such PRs into multiple smaller ones with stanadlone changes so as to be easier to test and review and ensure that  those are easily reversible if sth goes south (same for handling their backports etc).
I think you are right. Since the two changes are so related to each other (update the metrics - implies updating the test files as well - and moving the common code to just one file), I ended up committing them together. But I will be more careful next time.

@ChrsMark

File: metricbeat/module/kubernetes/_meta/test/docs/without-ksm/metricbeat-kubernetes.yaml (hunk 11 above)
Comment (constanca-m) [reply metricbeat/module/kubernetes/_meta/test/docs/without-ksm/metricbeat-kubernetes.yaml:180]: I must confess I did not check the files for testing. It was not the purpose of this PR, but I can check them later

**Comment by constanca-m** [comment]
I moved one of the main changes (move common code to just one file) to a new [PR](https://github.com/elastic/beats/pull/34432). I am leaving this one open for discussion.

**Comment by MichaelKatsoulis** [comment]
I believe that the goal of this PR should be to use the new KSM fields, but map them to the same elasticsearch fields as before. That way we will avoid breaking changes.
If in some cases it is not possible we need to be very careful with the impact, check if this field is used anywhere in a dashboard and think of a user's upgrade process.

**Comment by constanca-m** [comment]
> I believe that the goal of this PR should be to use the new KSM fields, but map them to the same elasticsearch fields as before. That way we will avoid breaking changes. If in some cases it is not possible we need to be very careful with the impact, check if this field is used anywhere in a dashboard and think of a user's upgrade process.

I think the name of the fields remains the same @MichaelKatsoulis, at least in the example mentioned by @ChrsMark . I only removed the deprecated metrics, and added the new ones.

**Comment by ChrsMark** [comment]
> > ```
> > 1. I see in the description that some fields are removed, for example:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > kube_pod_container_resource_limits_*_cores replaced by kube_pod_container_resource_limits
> > kube_pod_container_resource_requests_*_cores replaced by kube_pod_container_resource_requests
> > kube_pod_container_status_terminated removed since it can be easily obtained through
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is a breaking change and should be discussed if needed. In general we should not remove fields with no serious reason, since it would break users' dashboards, alerts, visualisations (even custom ones). Can we clarify this please here?
> 
> > ```
> > 2. I see that golden files (.expected) are changed along with the code's changes. This means that unit-tests' files are changed at the same time with the code's changes. How we ensure here that code changes do not break anything? In principle that's the purpose of the unit tests to ensure that code changes do not break the functionality and do not break backwards compatibility. Are we confident here that code changes are transparent to the end results our users will face?
> > ```
> 
> Since we only support Kubernetes versions that are still being maintained, it doesn't really make sense to keep checking metrics from Kube state metrics versions that are no longer compatible to what we support. We have been removing these kind of files in previous [PRs](https://github.com/elastic/beats/pull/32037/files#diff-2275d963a655d60e65554a9e35bccb30697c9f6c394c9711451701844f9dcc44).


I think I got it, thank you for elaborating. Could you update the PR description to make it more clear what fields (as we store them in ES) are removed or changed? As @MichaelKatsoulis already mentioned we need to make sure that fields that are stored in ES are the same with what we had in the past, no matter from what ksm metric we extract them for the recent versions. 

Also I'm not 100% confident of completely removing fields that are not supported/exposed by newest ksm versions. Maybe we can just leave them there for now and add a comment next to them in the code to remove them in the future. This would ensure us that even users that use older versions would still have some time to adopt. Not strong preference here, up to you to decide maybe with @gizas.


> 
> > ```
> > 3. Are dashboards also tested? Since these changes touch the ingest part we need to make sure that pre-built dashboards are also functional.
> > ```
> 
> They are not, I mentioned it in the PR description. I didn't check it because we still haven't established what we want to convey from them. I wanted to address that by writing [documentation in one of our resource metrics](https://github.com/elastic/observability-docs/issues/2474) (controller manager, in this case) so we could discuss what it is important and what is not. While we don't do that, I don't know if it is worth it changing so many dashboards to end up changing them later as they don't hold the meaning we want.
> 

Actually I'm not referring to updating any dashboards, just to run manual tests to verify that with this PR's changes the pre-built dashboards are still working without an issue. If we find out that the pre-built dashboards are not working because some fields/metrics are removed we need to note this down and make a plan to update them properly. So do we know if the pre-built dashboards are successfully populated with this current PR's changes?

> > ```
> > 4. In principle it's a good practice to split such PRs into multiple smaller ones with stanadlone changes so as to be easier to test and review and ensure that  those are easily reversible if sth goes south (same for handling their backports etc).
> > ```
> 
> I think you are right. Since the two changes are so related to each other (update the metrics - implies updating the test files as well - and moving the common code to just one file), I ended up committing them together. But I will be more careful next time.

Sounds good!
